app/models/course/assessment/category.rb

```diff
@@ -1,6 +1,31 @@
 # Represents a category of assessments. This is typically 'Mission' and 'Training'.
 class Course::Assessment::Category < ActiveRecord::Base
   belongs_to :course, inverse_of: :assessment_categories
-  has_many :tabs, class_name: Course::Assessment::Tab.name, inverse_of: :category
+  has_many :tabs, class_name: Course::Assessment::Tab.name,
+           inverse_of: :category, dependent: :destroy
   has_many :assessments, through: :tabs
+
+  accepts_nested_attributes_for :tabs
+
+  default_scope { order(:weight) }
+
+  before_destroy :before_destroy_validation
+
+  # Returns a boolean value indicating if this category can be safely destroyed.
+  # This is the case when another category with at least one tab exists.
+  #
+  # @return [Boolean]
+  def safe_to_destroy?
+    categories = course.assessment_categories.includes(:tabs)
+    other_tabs_remaining = categories.map(&:tabs).flatten.count - tabs.count
```

Comment:
why not just prevent a tab from being deleted if it's the last one, and then just the category checking it's not the last one?

I don't see why there's a need to check both categories and tabs, when the invariant never allows a category without a tab.